shared/src/balancer/pool_fetching.rs

```diff
@@ -0,0 +1,169 @@
+use anyhow::Result;
+use model::TokenPair;
+use std::collections::{HashMap, HashSet};
+
+use crate::balancer::event_handler::{PoolRegistry, PoolSpecialization, RegisteredPool};
+use crate::pool_fetching::{handle_contract_error, Block, MAX_BATCH_SIZE};
+use crate::Web3;
+use contracts::{BalancerV2Vault, BalancerV2WeightedPool};
+use ethcontract::batch::CallBatch;
+use ethcontract::errors::MethodError;
+use ethcontract::{BlockId, Bytes, H160, H256, U256};
+use itertools::Itertools;
+
+pub struct PoolTokenState {
+    pub balance: U256,
+    pub weight: U256,
+}
+
+pub struct WeightedPool {
+    pub pool_id: H256,
+    pub pool_address: H160,
+    pub specialization: PoolSpecialization,
+    pub reserves: HashMap<H160, PoolTokenState>,
+}
+
+impl WeightedPool {
+    fn new(pool_data: RegisteredPool, balances: Vec<U256>, weights: Vec<U256>) -> Self {
+        let mut reserves = HashMap::new();
+        // We expect the weight and token indices are aligned with balances returned from EVM query.
+        // If necessary we would also pass the tokens along with the query result,
+        // use them and fetch the weights from the registry by token address.
+        for (i, balance) in balances.into_iter().enumerate() {
+            reserves.insert(
+                pool_data.tokens[i],
+                PoolTokenState {
+                    balance,
+                    weight: weights[i],
+                },
+            );
+        }
+        WeightedPool {
+            pool_id: pool_data.pool_id,
+            pool_address: pool_data.pool_address,
+            specialization: pool_data.specialization,
+            reserves,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+pub trait WeightedPoolFetching: Send + Sync {
+    async fn fetch(
+        &self,
+        token_pairs: HashSet<TokenPair>,
+        at_block: Block,
+    ) -> Result<Vec<WeightedPool>>;
+}
+
+pub struct BalancerPoolFetcher {
+    pool_data: PoolRegistry,
+    vault: BalancerV2Vault,
+    web3: Web3,
+}
+
+#[async_trait::async_trait]
+impl WeightedPoolFetching for BalancerPoolFetcher {
+    async fn fetch(
+        &self,
+        token_pairs: HashSet<TokenPair>,
+        at_block: Block,
+    ) -> Result<Vec<WeightedPool>> {
+        let mut batch = CallBatch::new(self.web3.transport());
+        let block = BlockId::Number(at_block.into());
+        let futures = token_pairs
+            .into_iter()
+            .flat_map(|pair| self.pool_data.pools_containing_pair(pair))
+            .unique_by(|pool| pool.pool_id)
+            .map(|weighted_pool| {
```

Comment:
This still has the issue that we don't know that the pools here are, in fact, `WeightedPool`s and that they may be some other pool implementation.

Just because they have a `normalizedWeights` contract method doesn't mean that they will follow the logic from the weighted pool that we are expecting.

I added a comment in #696 mentioning this as well, but I think we should instead listen to events from the `WeightedPoolFactory` instead of the `Vault` for pool creation. This way we are guaranteed to have the weighted pools that we expect.

Reply:
Wouldn't it also be appropriate to simply filter out those events which are not in correspondence with weighted pools? Although I we would also solve the problem of having to filter out all the other events coming from the vault. I will rework the first implementation then.

Reply:
> This may mean that we have to make a web3 query to fetch the tokens

I think that its fine. This is also part of the "immutable" part of the the pool, so we can fetch it at the same time as we do the normalized weights and cache it permanently.

Reply:
No longer relevant as the weights are fetched and stored by our Event Handler.